### PR TITLE
feat(apps/whale-api): add estimated dex fees on get bestpath

### LINKED ISSUE #1574: include `DEX fees` on the response of `getBestPath`
#### What would you like to be added:
Related - https://github.com/DeFiCh/wallet/issues/2882

We want to display the DEX fees on the LW. Currently, LW uses `getBestPath` for every swap. We can include the calculation of `DEX fees` for it so it can be used on other client apps as well.

### PR DESCRIPTION
**Pull request by pierregee**
#### What this PR does / why we need it:
- Retrieve dex fee percentage from `gov attributes`
- Add estimated return and dex fees on each leg 
 

#### Which issue(s) does this PR fixes?:

Fixes #1574

#### Additional comments?:
- bidirectional swap path on tests won't work anymore because of of the added `estimatedReturn`

### CODE REVIEW

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 1]
```diff
@@ -255,8 +270,8 @@ export class PoolSwapPathFindingService {
   }
 }
 
-function computeReturnInDestinationToken (path: SwapPathPoolPair[], fromTokenId: string): BigNumber {
-  let total = new BigNumber(1)
+function computeReturnInDestinationToken (path: SwapPoolPair[], fromTokenId: string, previousTotal?: BigNumber): BigNumber {
+  let total = previousTotal === undefined ? new BigNumber(1) : previousTotal
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts:274]:
```suggestion
  let total = previousTotal === previousTotal ?? new BigNumber(1);
```

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 2]
```diff
@@ -129,9 +134,9 @@ export class PoolSwapPathFindingService {
 
     for (const path of allSimplePaths(this.tokenGraph, fromTokenId, toTokenId)) {
       const poolPairs: SwapPathPoolPair[] = []
-
       // Iterate over the path pairwise; ( tokenA )---< poolPairId >---( tokenB )
       // to collect poolPair info into the final result
+      let estimatedReturnTokenA = new BigNumber(1)
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts:139]: can move to ln: 163?? (read friendly)

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    protected readonly deFiDCache: DeFiDCache,
+    protected readonly cache: SemaphoreCache,
+    protected readonly poolSwapAggregatedMapper: PoolSwapAggregatedMapper,
+    protected readonly voutMapper: TransactionVoutMapper
+  ) {
+  }
+
+  public async getDexFees (tokenPathWithDirection: SwapPathWithDirection): Promise<EstimatedDexFee[] | []> {
+    const dexFees: EstimatedDexFee[] = []
+    const rpcResult = await this.rpcClient.masternode.getGov('ATTRIBUTES')
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts:25]: can move to ln: 28?? (read friendly)

Comment (canonbrother) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts:25]:
```suggestion
    const attrs = await this.rpcClient.masternode.getGov('ATTRIBUTES')['ATTRIBUTES']
```

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 4]
```diff
@@ -368,12 +368,14 @@ describe('get best path', () => {
...
-          tokenB: { id: '0', symbol: 'DFI', displaySymbol: 'DFI' }
+          tokenB: { id: '0', symbol: 'DFI', displaySymbol: 'DFI' },
+          estimatedReturn: '2.00000000',
+          estimatedDexFees: []
         }
       ],
       estimatedReturn: '2.00000000'
     })
-    expect(paths1.bestPath).toStrictEqual(paths2.bestPath)
+    // expect(paths1.bestPath).toStrictEqual(paths2.bestPath)
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts:378]: hmm.. shouldn't the `paths2` contain the estimated value?? :thinking:

Comment (eli-lim) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts:378]: `paths1.bestPath[0].estimatedReturn` value is denoted in DFI
`paths2.bestPath[0].estimatedReturn` value is denoted in A

I think we can delete this test - it might not always be the case that the best path from a->b will be the same as b->a

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts (hunk 1 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: Updated!
`let total = previousTotal ?? new BigNumber(1)`

File: apps/whale-api/src/module.api/poolpair.fees.service.ts (hunk 3 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.ts]: Updated!

Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.ts]: Updated! 
`    const attrs = (await this.rpcClient.masternode.getGov('ATTRIBUTES')).ATTRIBUTES`

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts (hunk 4 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.controller.e2e.ts]: Updated (removed)!

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 5]
```diff
@@ -23,6 +26,8 @@ export class PoolSwapPathFindingService {
 
   constructor (
     protected readonly deFiDCache: DeFiDCache,
+    private readonly poolPairFeesServices: PoolPairFeesService,
+
```
[the remark this answers is not in the document]
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: Updated! Removed extra line

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+@Injectable()
+export class PoolPairFeesService {
+  constructor (
+    @Inject('NETWORK') protected readonly network: NetworkName,
+    protected readonly rpcClient: JsonRpcClient,
+    protected readonly deFiDCache: DeFiDCache,
+    protected readonly cache: SemaphoreCache,
+    protected readonly poolSwapAggregatedMapper: PoolSwapAggregatedMapper,
+    protected readonly voutMapper: TransactionVoutMapper
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts]: I see deFiDCache, poolSwapAggregatedMapper and voutMapper aren't used - they should be removed 
except for deFiDCache (see my comment on line 27)

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  ) {
+  }
+
+  public async getDexFees (tokenPathWithDirection: SwapPathWithDirection): Promise<EstimatedDexFee[] | []> {
+    const dexFees: EstimatedDexFee[] = []
+    const { tokenFrom, tokenTo, poolPairId } = tokenPathWithDirection
+
+    const attrs = (await this.rpcClient.masternode.getGov('ATTRIBUTES')).ATTRIBUTES
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts]: Possible to rely on deFiDCache for this? Since the attributes seldom change, I think relying on cache will be more performant.

File: apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+  await setup()
+
+  app = await createTestingApp(container)
+  controller = app.get(PoolPairController)
+  const defiCache = app.get(DeFiDCache)
+
+  const tokenResult = await container.call('listtokens')
+  // precache
+  for (const k in tokenResult) {
+    await defiCache.getTokenInfo(k) as TokenInfo
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]:
```suggestion
    await defiCache.getTokenInfo(k)
```

unnecessary cast?

File: apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,363 @@
+import { PoolPairController } from './poolpair.controller'
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: `poolpair.fees.controller.e2e.ts` -> can we call it `poolpair.fees.service.e2e.ts`, since there's no PoolpairFeesController?

File: apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+    expect(paths1.bestPath).toStrictEqual([
+      {
+        estimatedDexFees: [{
+          amount: '0.00700000',
+          token: {
+            displaySymbol: 'dBTC',
+            symbol: 'BTC'
+          }
+        }],
+        estimatedReturn: '10.00000000',
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: is this denoted in dBTC or dDOGE?
I think it's possible to change the structure to something like

```typescript
bestPath: [
  {
    estimates: {
      dexFee: '0.00700000',
      return: '10.00000000'
      token: {
        displaySymbol: 'dBTC',
        symbol: 'BTC'
      }
    }
  }
]

Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: neither, it is DFI (tokenB)

 hmm the estimated dex fees should be an array because there could be a fee on both tokenA and tokenB and it's not useful to add an estimatedReturn on `tokenA` 
[Sample e2e with 2 dex fees on one poolpair](https://github.com/JellyfishSDK/jellyfish/pull/1601/files/85dc86b23c325fa4c27c00d4845ae05702d82f02#diff-7f86947ae35f08cfa839fbbd670a46f364d1edda7e2f4398b6b23e54a52cf1cbR293)

File: apps/whale-api/src/module.api/poolpair.fees.service.ts (hunk 7 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.ts]: Possible and updated! But, needed to combine -> split id to comply with `this.get`

File: apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts (hunk 8 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: yes updated!

File: apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts (hunk 10 above)
Comment (eli-lim) [reply apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: ah got it. so the `estimatedReturn` in the path is always denoted in tokenB (never tokenA)?

Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: yes, correct!

File: apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts (hunk 9 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.controller.e2e.ts]: oh, 🙈 updated!

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    const poolpairFee = attrs[`v0/poolpairs/${poolPairId}/token_a_fee_pct`]
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts]: IIRC, there could possibly be a `token_b_fee_pct`, so we'll have to account for that as well

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 12]
```diff
@@ -154,6 +158,16 @@ export class PoolSwapPathFindingService {
...
+        }
+
+        const estimatedReturnTokenB = computeReturnInDestinationToken([swapPathPoolPair], tokenA, estimatedReturnTokenA)
+        const swapPathWithDirection = getSwapPathWithDirection(swapPathPoolPair, tokenA, tokenB, estimatedReturnTokenA, estimatedReturnTokenB)
+        const poolpairDexFees = await this.poolPairFeesServices.getDexFees(swapPathWithDirection)
+        estimatedReturnTokenA = estimatedReturnTokenB
+        poolPairs.push({
+          ...swapPathPoolPair,
+          estimatedReturn: estimatedReturnTokenB.toFixed(8),
+          estimatedDexFees: poolpairDexFees
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: Noticed that `estimatedReturn` here is just a duplication of `priceRatio.ba`.
I suggest the following:

```typescript
{
  priceRatio: { ab: '...', ba: '...' },
  dexFees: { ab: '...', ba: '...' }
}
```

**Comment by fuxingloh** [comment]
I'm going to rebase this branch due to upstream docker changes.

@pierregee there were some updates to fee flow, leaving https://github.com/JellyfishSDK/jellyfish/pull/1577 here just in case you missed it.

**Review by surangap — COMMENTED** [review]
@eli-lim @pierregee 
Adding this here since `Asymmetric pool swap fee` came after discussion of this PR
please check the PRs -> https://github.com/JellyfishSDK/jellyfish/pull/1577
https://github.com/DeFiCh/ain/pull/1354
it might have an impact

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts (hunk 12 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: @eli-lim ab and ba kinda denotes that it is a path, I think it more makes sense to use
``` 
estimatedDexFeesInPct: {
          tokenA: '0.00300000',
          tokenB: '0.00000000'
        },
```

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+import { PoolPairInfo } from '@defichain/jellyfish-api-core/dist/category/poolpair'
+
+@Injectable()
+export class PoolPairFeesService {
+  constructor (
+    protected readonly rpcClient: JsonRpcClient
+  ) {
+  }
+
+  public async getDexFeesPct (poolPair: PoolPairInfo, poolPairId: string, fromToken: string, toToken: string): Promise<EstimatedDexFees | undefined> {
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts]: hmm.. i think can move this to `poolswap.pathfinding.service.ts` as a private function??

**Review by canonbrother — COMMENTED** [review]
.

File: apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+      'v0/poolpairs/6/token_a_fee_pct': '0.003', // DOG
+      'v0/poolpairs/6/token_a_fee_direction': 'out', // DOG
+      'v0/poolpairs/6/token_b_fee_pct': '0.004', // DFI
+      'v0/poolpairs/6/token_b_fee_direction': 'out', // DFI
+
+      'v0/poolpairs/8/token_a_fee_pct': '0.005', // KOALA
+      'v0/poolpairs/8/token_a_fee_direction': 'in', // KOALA
+
+      'v0/poolpairs/10/token_b_fee_pct': '0.006', // FISH
+      'v0/poolpairs/10/token_b_fee_direction': 'out' // FISH
```
Comment (canonbrother) [inline apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts]: here needs to test the `both` as well
eg:
```
'v0/poolpairs/{id}/token_b_fee_pct': '0.007',
'v0/poolpairs/{id}/token_b_fee_direction': 'both'
```

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 15]
```diff
@@ -350,7 +350,6 @@ describe('get', () => {
 describe('get best path', () => {
   it('should be bidirectional swap path - listPaths(a, b) === listPaths(b, a)', async () => {
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts:353]: can delete this test

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 16]
```diff
@@ -531,7 +528,6 @@ describe('get all paths', () => {
         ]
       ]
     })
-    expect(paths1.paths).toStrictEqual(paths2.paths)
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts:536]: I think this isn't necessary to delete?

File: packages/whale-api-client/src/api/poolpairs.ts [hunk 17]
```diff
@@ -243,6 +243,11 @@ export interface AllSwappableTokensResult {
   swappableTokens: TokenIdentifier[]
 }
 
+export interface EstimatedDexFees {
+  tokenA: string
+  tokenB: string
```
Comment (eli-lim) [inline packages/whale-api-client/src/api/poolpairs.ts]:
```suggestion
  tokenA: string // BigNumber
  tokenB: string // BigNumber
```

Thought these were referring to the token's name / symbol at first 😛

File: apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts (hunk 14 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts]: Good catch! Updated!

File: packages/whale-api-client/src/api/poolpairs.ts [hunk 18]
```diff
@@ -265,6 +270,7 @@ export interface SwapPathPoolPair {
     ab: string
     ba: string
   }
+  estimatedDexFees?: EstimatedDexFees
```
Comment (eli-lim) [inline packages/whale-api-client/src/api/poolpairs.ts]:
```suggestion
  dexFees?: DexFees
```

Looking at the implementation now, there's no actual "estimation" going on
Hence I think we can simplify the name

File: apps/whale-api/src/module.api/poolpair.fees.service.ts (hunk 13 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.ts]: @eli-lim what do you think? I rmb we had a discussion that it's better to be in another service CC: @canonbrother

Comment (eli-lim) [reply apps/whale-api/src/module.api/poolpair.fees.service.ts]: Hmm I think we can put all dexFee-related fetching / calculations in here so all the logic is consolidated in one place. Expose them as helpers methods to Poolswap.pathfinding.service and beyond.

File: packages/whale-api-client/src/api/poolpairs.ts [hunk 19]
```diff
@@ -243,11 +243,18 @@ export interface AllSwappableTokensResult {
   swappableTokens: TokenIdentifier[]
 }
 
+export interface EstimatedDexFeesInPct {
+  ab: string
+  ba: string
+
```
Comment (eli-lim) [inline packages/whale-api-client/src/api/poolpairs.ts]:
```suggestion
```

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+  }
+
+  public async getDexFeesPct (poolPair: PoolPairInfo, fromToken: string, toToken: string): Promise<EstimatedDexFeesInPct | undefined> {
+    const { dexFeeInPctTokenA, dexFeeOutPctTokenA, dexFeeInPctTokenB, dexFeeOutPctTokenB } = poolPair
+    const tokenADirection = poolPair.idTokenA === fromToken ? 'in' : 'out'
+    const tokenBDirection = poolPair.idTokenB === toToken ? 'out' : 'in'
+
+    if (dexFeeInPctTokenA === undefined && dexFeeOutPctTokenA === undefined && dexFeeInPctTokenB === undefined && dexFeeOutPctTokenB === undefined) {
+      return undefined
+    }
```
Comment (surangap) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts:21]: hmm. might have a problem here. asymmetric fee keys are not always set. specially in the old way, only the `token_a_fee_pct`,  `token_b_fee_pct` is set and it's considered as direction `both`. But this will not be updated in `token_a_fee_direction` or `token_b_fee_direction`. you can test this by tweaking this simple test -> https://github.com/JellyfishSDK/jellyfish/blob/a5afb411c106a6cf57bf4676bed0f849b896f6a9/packages/jellyfish-api-core/__tests__/category/masternode/setGov.test.ts#L352

File: apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts [hunk 21]
```diff
@@ -0,0 +1,510 @@
...
+      'v0/poolpairs/12/token_a_fee_pct': '0.007', // TURTLE
+      'v0/poolpairs/12/token_a_fee_direction': 'both', // TURTLE
+
+      'v0/poolpairs/14/token_b_fee_pct': '0.008', // PANDA
+      'v0/poolpairs/14/token_b_fee_direction': 'both', // PANDA
+
+      'v0/poolpairs/16/token_a_fee_pct': '0.009', // RABBIT
+      'v0/poolpairs/16/token_a_fee_direction': 'both', // RABBIT
+      'v0/poolpairs/16/token_b_fee_pct': '0.010', // RABBIT
+      'v0/poolpairs/16/token_b_fee_direction': 'both' // RABBIT
```
Comment (surangap) [inline apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts]: add another pair and a test where only `token_a_fee_pct`, `token_b_fee_pct` is set.
related to https://github.com/JellyfishSDK/jellyfish/pull/1601/files#r919629537

File: apps/whale-api/src/module.api/poolpair.fees.service.ts (hunk 20 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.ts:21]: Thanks for the explanation, @surangap! Tested this scenario and able to validate that there's no changes needed in this line because `dexFeeInPct**` is populated if direction is `not set` in `gov`

Regardless if there's no change needed here, I still added [e2e tests ](https://github.com/JellyfishSDK/jellyfish/pull/1601/commits/db69caf13d74c127498d9e663cac9f6b36911eee#diff-d0967f195ebd61af3cfd8c80b467983958ae965caa4857a2d283b4095b6cb07eR498)wherein direction is `not set` (considered as `both`)

File: apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts (hunk 21 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolpair.fees.service.e2e.ts]: [Here](https://github.com/JellyfishSDK/jellyfish/pull/1601/commits/db69caf13d74c127498d9e663cac9f6b36911eee#diff-d0967f195ebd61af3cfd8c80b467983958ae965caa4857a2d283b4095b6cb07eR498), good catch, @surangap! 🙏

**Review by canonbrother — DISMISSED** [review]

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 22]
```diff
@@ -58,20 +60,27 @@ export class PoolSwapPathFindingService {
           toToken: toToken,
           bestPath: path,
           estimatedReturn: formatNumber(
-            computeReturnInDestinationToken(path, fromTokenId)
+            computeReturnLessDexFeesInDestinationToken(path, fromTokenId).estimatedReturn
+          ), // denoted in toToken
+          estimatedReturnLessDexFees: formatNumber(
+            computeReturnLessDexFeesInDestinationToken(path, fromTokenId).estimatedReturnLessDexFees
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: `computeReturnLessDexFeesInDestinationToken` can just be called once here instead of twice

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 23]
```diff
@@ -58,20 +60,27 @@ export class PoolSwapPathFindingService {
...
     let bestPath: SwapPathPoolPair[] = []
+    let bestReturnLessDexFees = new BigNumber(0)
     let bestReturn = new BigNumber(0)
 
     for (const path of paths) {
-      const totalReturn = computeReturnInDestinationToken(path, fromTokenId)
-      if (totalReturn.isGreaterThan(bestReturn)) {
-        bestReturn = totalReturn
+      const { estimatedReturnLessDexFees, estimatedReturn } = computeReturnLessDexFeesInDestinationToken(path, fromTokenId)
+      if (estimatedReturn.isGreaterThan(bestReturnLessDexFees)) {
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]:
```suggestion
      if (estimatedReturn.isGreaterThan(bestReturn)) {
```

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts [hunk 24]
```diff
@@ -255,18 +268,45 @@ export class PoolSwapPathFindingService {
...
       fromTokenId = poolPair.tokenA.id
+      priceRatio = poolPair.priceRatio.ab
+      fromTokenFeePct = poolPair.estimatedDexFeesInPct?.ab
+      toTokenFeePct = poolPair.estimatedDexFeesInPct?.ba
     }
+
+    estimatedReturn = estimatedReturn.multipliedBy(priceRatio)
+
+    // less dex fee token A
+    const estimatedDexFeeInTokenA = new BigNumber(fromTokenFeePct ?? 0).multipliedBy(estimatedReturnLessDexFees)
```
Comment (eli-lim) [inline apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: Correct me if I'm wrong, but I think `estimatedDexFeeInTokenA` here is misleading as `fromToken` could refer to either token A or B (not just A) right?

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts (hunk 23 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: Updated!

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts (hunk 22 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: Updated!

File: packages/whale-api-client/src/api/poolpairs.ts (hunk 19 above)
Comment (pierregee) [reply packages/whale-api-client/src/api/poolpairs.ts]: Updated!

File: apps/whale-api/src/module.api/poolswap.pathfinding.service.ts (hunk 24 above)
Comment (pierregee) [reply apps/whale-api/src/module.api/poolswap.pathfinding.service.ts]: You're right, it's quite misleading and better to use fromToken, updated!

**Review by surangap — APPROVED** [review]

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+import BigNumber from 'bignumber.js'
+import { EstimatedDexFeesInPct } from '@defichain/whale-api-client/dist/api/poolpairs'
+import { PoolPairInfo } from '@defichain/jellyfish-api-core/dist/category/poolpair'
+
+@Injectable()
+export class PoolPairFeesService {
+  constructor (
+    protected readonly rpcClient: JsonRpcClient
+  ) {
+  }
```
Comment (fuxingloh) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts:12]:
```suggestion
```

File: apps/whale-api/src/module.api/poolpair.fees.service.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
+import { Injectable } from '@nestjs/common'
+import { JsonRpcClient } from '@defichain/jellyfish-api-jsonrpc'
```
Comment (fuxingloh) [inline apps/whale-api/src/module.api/poolpair.fees.service.ts:2]:
```suggestion
```

File: apps/whale-api/src/module.api/poolpair.controller.e2e.ts [hunk 27]
```diff
@@ -373,9 +372,9 @@ describe('get best path', () => {
           tokenB: { id: '0', symbol: 'DFI', displaySymbol: 'DFI' }
         }
       ],
-      estimatedReturn: '2.00000000'
+      estimatedReturn: '2.00000000',
+      estimatedReturnLessDexFees: '2.00000000'
     })
-    expect(paths1.bestPath).toStrictEqual(paths2.bestPath)
```
Comment (fuxingloh) [inline apps/whale-api/src/module.api/poolpair.controller.e2e.ts:378]: Why was this removed?

**Review by eli-lim — APPROVED** [review]
:shipit:

**Review by canonbrother — APPROVED** [review]

**Review by fuxingloh — CHANGES_REQUESTED** [review]
Should add check for `tradeEnabled` at isPoolPairIgnored()

**Comment by fuxingloh** [comment]
Ignore my comments and resolve in another PR, I will merge this in after the test passes
